lib/semantics/expression.h

```diff
@@ -27,30 +29,100 @@ template<typename> struct Integer;
+// The ExpressionAnalysisContext wraps a SemanticsContext reference
+// and implements constraint checking on expressions using the
+// parse tree node wrappers that mirror the grammar annotations used
+// in the Fortran standard (i.e., scalar-, constant-, &c.).  These
+// constraint checks are performed in a deferred manner so that any
+// errors are reported on the most accurate source location available.
 
-class SemanticsContext;
+namespace Fortran::evaluate {
+class ExpressionAnalysisContext {
+public:
+  using ConstraintChecker = bool (ExpressionAnalysisContext::*)(
+      Expr<SomeType> &);
 
-// Semantic analysis of one expression.
-std::optional<evaluate::Expr<evaluate::SomeType>> AnalyzeExpr(
-    SemanticsContext &, const parser::Expr &);
+  ExpressionAnalysisContext(semantics::SemanticsContext &sc) : context_{sc} {}
+  ExpressionAnalysisContext(ExpressionAnalysisContext &i)
+    : context_{i.context_}, inner_{&i} {}
+  ExpressionAnalysisContext(ExpressionAnalysisContext &i, ConstraintChecker cc)
+    : context_{i.context_}, inner_{&i}, constraint_{cc} {}
 
+  semantics::SemanticsContext &context() const { return context_; }
+
+  template<typename... A> void Say(A... args) {
+    context_.foldingContext().messages.Say(std::forward<A>(args)...);
+  }
+
+  void CheckConstraints(std::optional<Expr<SomeType>> &);
+  bool ScalarConstraint(Expr<SomeType> &);
+  bool ConstantConstraint(Expr<SomeType> &);
+  bool IntegerConstraint(Expr<SomeType> &);
+
+protected:
+  semantics::SemanticsContext &context_;
+
+private:
+  ExpressionAnalysisContext *inner_{nullptr};
+  ConstraintChecker constraint_{nullptr};
+};
+
+template<typename PARSED>
+std::optional<Expr<SomeType>> AnalyzeExpr(
+    ExpressionAnalysisContext &, const PARSED &);
+
+// This extern template is the gateway into the rest of the expression
+// analysis implementation in expression.cc.
+extern template std::optional<Expr<SomeType>> AnalyzeExpr(
+    ExpressionAnalysisContext &, const parser::Expr &);
+
+// These specializations create nested expression analysis contexts
+// to implement constraint checking.
+
+template<typename A>
+std::optional<Expr<SomeType>> AnalyzeExpr(
+    ExpressionAnalysisContext &context, const parser::Scalar<A> &expr) {
+  ExpressionAnalysisContext withCheck{
+      context, &ExpressionAnalysisContext::ScalarConstraint};
+  return AnalyzeExpr(withCheck, expr.thing);
+}
+
+template<typename A>
+std::optional<Expr<SomeType>> AnalyzeExpr(
+    ExpressionAnalysisContext &context, const parser::Constant<A> &expr) {
+  ExpressionAnalysisContext withCheck{
+      context, &ExpressionAnalysisContext::ConstantConstraint};
+  return AnalyzeExpr(withCheck, expr.thing);
+}
+
+template<typename A>
+std::optional<Expr<SomeType>> AnalyzeExpr(
+    ExpressionAnalysisContext &context, const parser::Integer<A> &expr) {
+  ExpressionAnalysisContext withCheck{
+      context, &ExpressionAnalysisContext::ConstantConstraint};
+  return AnalyzeExpr(withCheck, expr.thing);
+}
```

Comment:
I think we also need `AnalyzeExpr(context, const common::Indirection<A> &expr)` to handle types like `parser::ScalarIntConstantExpr` that have `Indirection` in them.

Reply:
Ok, will hoist that one out to the header file as well.